lib/SPIRV/SPIRVBuiltinHelper.cpp

```diff
@@ -179,30 +185,21 @@ BuiltinCallMutator &BuiltinCallMutator::replaceArg(unsigned Index,
   Args[Index] = Arg.first;
   PointerTypes[Index] = Arg.second;
   Attrs = Attrs.removeParamAttributes(CI->getContext(), Index);
+  CallAttrs = CallAttrs.removeParamAttributes(CI->getContext(), Index);
   return *this;
 }
 
 BuiltinCallMutator &BuiltinCallMutator::removeArg(unsigned Index) {
   // If the argument being dropped is the last one, there is nothing to move, so
   // just remove the attributes.
+  auto &Ctx = CI->getContext();
   if (Index == Args.size() - 1) {
-    // TODO: Remove this workaround when LLVM fixes
-    // https://github.com/llvm/llvm-project/issues/59746 on
-    // AttributeList::removeParamAttributes function.
-    // AttributeList::removeParamAttributes function sets attribute at
-    // specified index empty so that return value of
-    // AttributeList::getNumAttrSet() keeps unchanged after that call. When call
-    // BuiltinCallMutator::removeArg function, there is assert failure on
-    // BuiltinCallMutator::doConversion() since new CallInst removed arg but
-    // still holds attribute of that removed arg.
-    SmallVector<AttributeSet, 4> ArgAttrs;
-    for (unsigned I = 0; I < Index; ++I)
-      ArgAttrs.push_back(Attrs.getParamAttrs(I));
-    Attrs = AttributeList::get(CI->getContext(), Attrs.getFnAttrs(),
-                               Attrs.getRetAttrs(), ArgAttrs);
-  } else
-    moveAttributes(CI->getContext(), Attrs, Index + 1, Args.size() - Index - 1,
-                   Index);
+    Attrs = Attrs.removeParamAttributes(Ctx, Index);
+    CallAttrs = CallAttrs.removeParamAttributes(Ctx, Index);
+  } else {
+    moveAttributes(Ctx, Attrs, Index + 1, Args.size() - Index - 1, Index);
```

Comment:
I'm not quite following, so call instruction and function declaration have different sets of attributes? Is it just about tail call?

Reply:
Yes, they could be different.
As shown in the input IR in PR description:
The callInst has a `noundef` attribute for return value, and each argument also has `noundef` attribute. So the callinst has 3 AttributeSets.
On the other hand, the called function _Z42__spirv_GenericCastToPtrExplicit_ToPrivatePvi declaration doesn't have these 3 attributes.

Tail call isn't a part of callinst attributes. `NewCall->setTailCall(CI->isTailCall());` is added in this PR to ensure tail call isn't lost.

Original `Attrs` is only used for creating new function at https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/1088aca0fd5ed0b855881bb52a199497dc86e952/lib/SPIRV/SPIRVUtil.cpp#L289